### PR TITLE
Add Matrix support

### PR DESCRIPTION
**Pull request by Jack-Works**
https://github.com/DimensionDev/Maskbook/issues/168

https://github.com/DimensionDev/Maskbook/issues/185

https://github.com/DimensionDev/Maskbook/issues/142

### CODE REVIEW

File: src/network/matrix/matrix.ts [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    private async lookupRoomAlias(alias: string): Promise<string> {
+        if (this.cache.has(alias)) return this.cache.get(alias)!
+        // @ts-ignore
+        const x: { room_id: string; servers: string[] } = await this.matrixClient.getRoomIdForAlias(alias)
+        this.cache.set(alias, x.room_id)
+        this.reverseCache.set(x.room_id, alias)
+        return x.room_id
+    }
+}
+export async function createMatrixMessage(username: string, password: string) {
```
Comment (guanbinrui) [inline src/network/matrix/matrix.ts]: Refactor these startup code as unit testing cases?

File: src/network/matrix/matrix.ts [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        unsigned: unknown & {
+            age: number
+        }
+        event_id: string
+        room_id: string
+    }
+}
+export class MatrixMessage {
+    public userID: null | string = null
+    constructor(public matrixClient: MatrixClient) {
```
Comment (guanbinrui) [inline src/network/matrix/matrix.ts]: Async logic in the `constructor` is not good.

Comment (Jack-Works) [reply src/network/matrix/matrix.ts]: If the client is not started the state might be mrssed

File: src/network/matrix/matrix.ts (hunk 1 above)
Comment (Jack-Works) [reply src/network/matrix/matrix.ts]: I didn't catch you r meaning

Comment (guanbinrui) [reply src/network/matrix/matrix.ts]: It looks just some test code here.

Comment (guanbinrui) [reply src/network/matrix/matrix.ts]: Never mind. Too many things need to mock here.

File: src/network/matrix/matrix.ts [hunk 3]
```diff
@@ -0,0 +1,175 @@
+/// <reference path="./matrix.type.d.ts" />
+import sdk, { MatrixClient } from 'matrix-js-sdk'
+import mitt from 'mitt'
+export const endpoint = 'https://matrix.vampire.rip'
```
Comment (SunriseFox) [inline src/network/matrix/matrix.ts]: better not use this...

Comment (Jack-Works) [reply src/network/matrix/matrix.ts]: We'll change to other url when go to production

**Pull request merged**